### PR TITLE
🌱 Add live ISO e2e test

### PR DESCRIPTION
**Pull request by adilGhaffarDev**
What this PR does / why we need it:
e2e test with live iso.

Discussion on slack regarding live iso:
https://kubernetes.slack.com/archives/CHD49TLE7/p1653654187046809

### CODE REVIEW

**Comment by adilGhaffarDev** [comment]
test-v1b1-e2e-ubuntu

File: scripts/environment.sh
Comment (mboukhalfa) [inline scripts/environment.sh:86]: I would prefer this to be in e2e-config file since it is needed only by the e2e. I am usually export vars here only if it is necessary but if the var is used only by e2e framework would be better to be in the e2e file

File: test/e2e/live_iso_test.go
Comment (mboukhalfa) [inline test/e2e/live_iso_test.go:17]: I think only 1 worker should be enough or even spinning only the kcp

File: test/e2e/live_iso_test.go
Comment (mboukhalfa) [inline test/e2e/live_iso_test.go:44]: add check here by reading `/var/log/libvirt/qemu/"{{ node name }}-serial0.log` and verify it contains `"#  Welcome" `

File: test/e2e/live_iso_test.go
Comment (mboukhalfa) [inline test/e2e/live_iso_test.go:66]: Since you are only verifying the kcp machine is running I believe we can simplify the condition and remove the fuction isWorker as follow:
```suggestion
		Eventually(func() (bool) {
			machineList := &clusterv1alpha4.MachineList{}
			if err := bootstrapClusterProxy.GetClient().List(ctx, machineList, client.InNamespace(namespace), client.MatchingLabels{clusterv1.ClusterLabelName: clusterName}); err == nil {
				for _, machine := range machineList.Items {
					if strings.EqualFold(machine.Status.Phase, "running") {
						return true
					}
				}
			}
			return false
		}, e2eConfig.GetIntervals(specName, "wait-worker-nodes")...).To(Succeed())
```

File: test/e2e/live_iso_test.go
Comment (mboukhalfa) [inline test/e2e/live_iso_test.go:80]: Why sleeping here ? can replace that directly with a loop check is the logs reached what we are expecting ?

**Review by mboukhalfa — COMMENTED** [review]
looks good check some suggestions

File: test/e2e/live_iso_test.go
Comment (adilGhaffarDev) [reply test/e2e/live_iso_test.go:66]: I tried the test by removing the check for worker node, but it started failing when I tried to get worker node after this loop, so I think we have to keep it there. Let me know if it can be improved in some other way.

File: Makefile
Comment (mboukhalfa) [inline Makefile:165]: /hold we cannot merge that else the other tests will not get triggered since ginkgo here focusing only on live-iso without condition, we should wait until the jjb https://gerrit.nordix.org/c/infra/cicd/+/13707 get merged then remove this line

File: test/e2e/live_iso_test.go
Comment (mboukhalfa) [reply test/e2e/live_iso_test.go:66]: can you please describe what was the issue ?

File: test/e2e/live_iso_test.go
Comment (lentzi90) [inline test/e2e/live_iso_test.go:52]:
```suggestion
		By("Waiting for the control plane and worker metal3machines to exist")
```

**Review by lentzi90 — COMMENTED** [review]
What is the reason we involve a full target cluster here? If I understand correctly we only check that the BMH can boot from the live-iso. Not that it can form a cluster (it probably cannot?). So I'm wondering about what the goal is.

Do we want to make sure that live-iso an be used also on the CAPM3 side (not just BMO) or was it just the easiest way to implement it since it is similar to other tests? It would be good to add some comment about this I think. The way it is now it can be quite confusing when seeing IP pools, metal3machines and clusters if we only check the boot log.

Alternatively, we could just apply/patch a BMH to use live-iso to boot and be done with it.

File: test/e2e/live_iso_test.go
Comment (adilGhaffarDev) [reply test/e2e/live_iso_test.go:52]: I would like to have @Xenwar input on this. To my knowledge for this test in cluster we need at least one node to be using live iso.

Comment (Xenwar) [reply test/e2e/live_iso_test.go]: I agree with @lentzi90 on
`Alternatively, we could just apply/patch a BMH to use live-iso to boot and be done with it.`
But, does the e2e code allows us to just provisioning a node, without setting up a cluster ?

Comment (lentzi90) [reply test/e2e/live_iso_test.go]: I think it should be possible. What I can see now in the live-iso test we start by creating a target cluster. So this means the bootstrap cluster is already there and it should also have the BMHs in available state.
Now we should be able to manipulate the BMH in a similar way like it is done in the remediation test where they [add annotations](https://github.com/metal3-io/cluster-api-provider-metal3/blob/548790c0acba5c10cdaf537e1e7fdbd9e80501ca/test/e2e/common.go#L109).

File: test/e2e/live_iso_test.go
Comment (lentzi90) [inline test/e2e/live_iso_test.go:36]: I think we can simplify this a bit.
```suggestion
		isoBmh := bmhs[0]
```
Maybe add an Èxpect` to check that it is not `nil`. But the point is that we don't have to fetch the BMH twice.

**Review by lentzi90 — APPROVED** [review]
Great job!
One small nit below, otherwise I'm happy with this.

**Review by lentzi90 — APPROVED** [review]
LGTM!

**Review by furkatgofurov7 — COMMENTED** [review]
@adilGhaffarDev can you squash your commits before merging?

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:20]: could you add a log indicating live ISO tests starting to keep it consistent with other tests. You can check this as an example: https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/pivoting_test.go#L33

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:47]:
```suggestion
		By("waiting for live ISO image booted host to be in provisioned state")
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:51]: I would suggest to use: `wait-object-provisioned`
```suggestion
		}, e2eConfig.GetIntervals(specName, "wait-object-provisioned")...).Should(Succeed())
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:56]:
```suggestion
		By("Reading serial logs to verify the node was booted from live ISO image")
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:22]: I would stick to capitalizing the acronym (ISO) in the code as it is a best practice in Go. 
```suggestion
		liveISOImageURL = e2eConfig.GetVariable("LIVE_ISO_IMAGE")
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:39]:
```suggestion
			URL:          liveISOImageURL,
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:33]:
```suggestion
		Expect(err).NotTo(HaveOccurred(), "Error getting BMHs")
```

File: test/e2e/live_iso_test.go
Comment (furkatgofurov7) [inline test/e2e/live_iso_test.go:62]: using `wait-worker-nodes` is confusing here, is not it? how about using: `wait-service` or `wait-job` if the task can be completed in 10 minutes?

**Review by furkatgofurov7 — CHANGES_REQUESTED** [review]
This is great, few suggestions inline:

**Comment by furkatgofurov7** [comment]
> test-v1b1-ubuntu-e2e-parallel

this is not triggered for some reason. 

/test-v1b1-ubuntu-e2e-parallel

**Comment by furkatgofurov7** [comment]
/retitle 🌱 Add live ISO e2e test

**Comment by kashifest** [comment]
@dtantsur since this test was requested by you, and its passing all the tests , if its looks good to you just unhold the PR.

/hold

**Comment by kashifest** [comment]
/hold cancel
Taking this in since there was no comment/objection from @dtantsur

**Comment by dtantsur** [comment]
Late to the party, but looks good to me, thanks!

**Pull request merged**
